process/block/processedMb/processedMiniBlocks.go

```diff
@@ -0,0 +1,135 @@
+
+// MiniBlockHashes will keep a list of miniblock hashes as keys in a map for easy access
+type MiniBlockHashes map[string]struct{}
+
+// ProcessedMiniBlockTracker is used to store all processed mini blocks hashes grouped by a metahash
+type ProcessedMiniBlockTracker struct {
+	processedMiniBlocks    map[string]MiniBlockHashes
+	mutProcessedMiniBlocks sync.RWMutex
+}
+
+// NewProcessedMiniBlocks will create a complex type of processedMb
+func NewProcessedMiniBlocks() *ProcessedMiniBlockTracker {
+	return &ProcessedMiniBlockTracker{
+		processedMiniBlocks: make(map[string]MiniBlockHashes),
+	}
+}
+
+// AddMiniBlockHash will add a miniblock hash
+func (pmb *ProcessedMiniBlockTracker) AddMiniBlockHash(metaBlockHash string, miniBlockHash string) {
+	pmb.mutProcessedMiniBlocks.Lock()
+	defer pmb.mutProcessedMiniBlocks.Unlock()
+
+	miniBlocksProcessed, ok := pmb.processedMiniBlocks[metaBlockHash]
+	if !ok {
+		miniBlocksProcessed := make(MiniBlockHashes)
+		miniBlocksProcessed[miniBlockHash] = struct{}{}
+		pmb.processedMiniBlocks[metaBlockHash] = miniBlocksProcessed
+
+		return
+	}
+
+	miniBlocksProcessed[miniBlockHash] = struct{}{}
+}
+
+// RemoveMetaBlockHash will remove a meta block hash
+func (pmb *ProcessedMiniBlockTracker) RemoveMetaBlockHash(metaBlockHash string) {
+	pmb.mutProcessedMiniBlocks.Lock()
+	delete(pmb.processedMiniBlocks, metaBlockHash)
+	pmb.mutProcessedMiniBlocks.Unlock()
+}
+
+// RemoveMiniBlockHash will remove a mini block hash
+func (pmb *ProcessedMiniBlockTracker) RemoveMiniBlockHash(miniBlockHash string) {
+	pmb.mutProcessedMiniBlocks.Lock()
+	for metaHash, miniBlocksProcessed := range pmb.processedMiniBlocks {
+		_, isProcessed := miniBlocksProcessed[miniBlockHash]
+		if isProcessed {
+			delete(miniBlocksProcessed, miniBlockHash)
+		}
+
+		if len(miniBlocksProcessed) == 0 {
+			delete(pmb.processedMiniBlocks, metaHash)
+		}
+	}
+	pmb.mutProcessedMiniBlocks.Unlock()
+}
+
+// GetProcessedMiniBlocksHashes will return all processed miniblocks for a metablock
+func (pmb *ProcessedMiniBlockTracker) GetProcessedMiniBlocksHashes(metaBlockHash string) map[string]struct{} {
+	pmb.mutProcessedMiniBlocks.RLock()
+	processedMiniBlocksHashes := pmb.processedMiniBlocks[metaBlockHash]
+	pmb.mutProcessedMiniBlocks.RUnlock()
+
+	return processedMiniBlocksHashes
+}
+
+// IsMiniBlockProcessed will return true if a mini block is processed
+func (pmb *ProcessedMiniBlockTracker) IsMiniBlockProcessed(metaBlockHash string, miniBlockHash string) bool {
+	pmb.mutProcessedMiniBlocks.RLock()
+	defer pmb.mutProcessedMiniBlocks.RUnlock()
+
+	miniBlocksProcessed, ok := pmb.processedMiniBlocks[metaBlockHash]
+	if !ok {
+		return false
+	}
+
+	_, isProcessed := miniBlocksProcessed[miniBlockHash]
+	return isProcessed
+}
+
+// ConvertProcessedMiniBlocksMapToSlice will convert a map[string]map[string]struct{} in a slice of MiniBlocksInMeta
+func (pmb *ProcessedMiniBlockTracker) ConvertProcessedMiniBlocksMapToSlice() []bootstrapStorage.MiniBlocksInMeta {
```

Comment:
I don't think this data structure should know about a boot storer. I think this conversion should be done by the boot storer which is the one who knows how it needs the data formatted

Reply:
i will let it as it is because if I move this method in another file I will need a mutex to protect  ProcessedMiniBlockTracker from concurrency accesses.